### PR TITLE
[refactor] Modify loss meter and relative logger

### PR DESCRIPTION
**Pull request by JanLin0817**

### CODE REVIEW

File: configs/pytorch_trainer/hook/trainer_hook.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+_BASE_: ./logger/logger_hook.yaml
+HOOK:
```
Comment (mousyball) [inline configs/pytorch_trainer/hook/trainer_hook.yaml:3]: Space after \_BASE\_

File: configs/pytorch_trainer/hook/logger/logger_hook.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+LOGGER_HOOK:
+  NAME:
+    - LossLoggerHook
+    - TensorboardLoggerHook
+    - TextLoggerHook
+
+  LossLoggerHook:
```
Comment (mousyball) [inline configs/pytorch_trainer/hook/logger/logger_hook.yaml:6]: Space among sub-fields for uniform behavior.

File: configs/pytorch_trainer/hook/trainer_hook.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    priority: NORMAL
+    interval: -1
+    out_dir:
+    save_meta: False
+    save_optimizer: True
+
+  SchedulerHook:
+    priority: VERY_LOW
+    mode: 'default'
+    interval: 1
```
Comment (mousyball) [inline configs/pytorch_trainer/hook/trainer_hook.yaml]: Newline at the end of file.

File: docs/priority.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# priority
+
+## priority table
+
+* The priority table is designed for controlling hook execution order
+    * HIGHEST means execution first
+* HIGHEST, VERY_HIGH, HIGH: hooks which requests data from the trainer before data been overwrite
+* NORMAL: orderless hook
+* LOW, VERY_LOW, LOWEST: hooks which changes the trainer data
```
Comment (mousyball) [inline docs/priority.md:10]: Rewrite the description to make it clear to understand.

File: examples/train_cifar10.py [hunk 5]
```diff
@@ -35,19 +36,17 @@ def train_step(self, batch_data):
 
         # forward
         outputs = self(inputs)
-        loss = self.criterion(outputs, labels)
+        losses = self.criterion(outputs, labels)
 
-        return dict(loss=loss,
-                    multi_loss=dict(cls_loss=loss))
+        return dict(cls_loss=losses)
```
Comment (mousyball) [inline examples/train_cifar10.py]: Reason of change?

File: pytorch_trainer/trainer/base_trainer.py [hunk 6]
```diff
@@ -50,18 +51,21 @@ def __init__(self,
...
+            date = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
+            work_dir = osp.join(work_dir, '')[:-1] + f'_{date}'
             self.work_dir = osp.abspath(work_dir)
             if not osp.isdir(work_dir):
                 os.makedirs(work_dir)
         elif work_dir is None:
             self.work_dir = None
         else:
-            raise TypeError('Argument "work_dir" must be a str or None')
+            raise TypeError('Argument "work_dir" must be a string')
```
Comment (mousyball) [inline pytorch_trainer/trainer/base_trainer.py]: end dot

File: pytorch_trainer/trainer/base_trainer.py [hunk 7]
```diff
@@ -50,18 +51,21 @@ def __init__(self,
         self.model = model
         self.optimizer = optimizer
         self.scheduler = scheduler
-        if logger is None:
-            self.logger = get_logger()
 
         # create work_dir
         if isinstance(work_dir, str):
+            date = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
+            work_dir = osp.join(work_dir, '')[:-1] + f'_{date}'
```
Comment (mousyball) [inline pytorch_trainer/trainer/base_trainer.py:58]: Example plz?
Does this handle the situation with or without slash?

File: pytorch_trainer/trainer/base_trainer.py [hunk 8]
```diff
@@ -98,6 +102,22 @@ def epoch(self):
     def max_epoch(self):
         return self._max_epoch
 
+    def _loss_parser(self, output):
+        """sum losses in output
```
Comment (mousyball) [inline pytorch_trainer/trainer/base_trainer.py]: Sum up the losses of output.

File: pytorch_trainer/trainer/hooks/logger/base_logger.py [hunk 9]
```diff
@@ -12,46 +12,20 @@ def __init__(self,
                  ):
```
Comment (mousyball) [inline pytorch_trainer/trainer/hooks/logger/base_logger.py]: Remove comma and follow the style.

```python
def __init__(self,
             interval=1):
```

File: pytorch_trainer/trainer/hooks/logger/loss_meter.py [hunk 10]
```diff
@@ -10,24 +10,27 @@ class LossLoggerHook(LoggerHook):
     LossLoggerHook only update and clear trainer loss_meters
```
Comment (mousyball) [inline pytorch_trainer/trainer/hooks/logger/loss_meter.py]: To make loss_meters independent ~~between Hook.~~ among hooks.

=> We did change the logic of loss meters. The hooks own their individual loss meter; Hence, need to revise the comment?

File: pytorch_trainer/trainer/hooks/logger/tensorboard.py [hunk 11]
```diff
@@ -39,42 +39,42 @@ def after_train_epoch(self, trainer):
...
-        for key, val in loss_dict.items():
-            self.writer.add_scalar('Train/{0}_epoch'.format(key), val, step)
-
-        lr_dict = self.get_lr_log(trainer)
-        for key, val in lr_dict.items():
-            self.writer.add_scalar('LR/{0}_epoch'.format(key), val, step)
+        self._add_loss_scalar(trainer, step)
+        self._add_lr_scalar(trainer, step)
 
     def after_train_batch(self, trainer):
```
Comment (mousyball) [inline pytorch_trainer/trainer/hooks/logger/tensorboard.py:45]: Rename `batch` to something else.

File: tests/trainer/hook_test.py [hunk 12]
```diff
@@ -10,83 +11,189 @@
...
+
+        assert osp.basename(glob(output_path)[0]) == expected
+
+        shutil.rmtree(work_dir)
+        shutil.rmtree(trainer.work_dir)
+
+    # TODO: logger hook test
 
 
 class Model(nn.Module):
```
Comment (mousyball) [inline tests/trainer/hook_test.py]: Extract test model to a single file.

File: tests/trainer/trainer_test.py [hunk 13]
```diff
@@ -1,51 +1,142 @@
...
+        return dict(loss=self(x)[0, 0])
+        # return dict(loss=self(x))
 
     def val_step(self, x):
-        return dict(loss=self(x))
+        return dict(loss=self(x)[0, 0])
+        # return dict(loss=self(x))
+
+
+class model_wrapper:
```
Comment (mousyball) [inline tests/trainer/trainer_test.py:36]: Mocking the situation of parallel module.

File: tests/trainer/trainer_test.py [hunk 14]
```diff
@@ -1,51 +1,142 @@
...
+        config.HOOK.OptimizerHook.update(cfg)
 
-        cfg = config()
-        trainer.register_callback(cfg)
+        # trainer
+        trainer = EpochBasedTrainer('model')
+        trainer._register_hook(config.HOOK)
         for hook in trainer._hooks:
             if isinstance(hook, OptimizerHook):
                 assert True and len(trainer._hooks) == 1
```
Comment (mousyball) [inline tests/trainer/trainer_test.py]: assert True then assert False
What is this test for?
It always triggers the assertion.

File: tests/trainer/trainer_test.py [hunk 15]
```diff
@@ -55,53 +146,152 @@ def test_regist_optimizer_hook(self, cfg):
...
+        # trainer
+        trainer = EpochBasedTrainer('model',
                                     scheduler='not_none')
 
-        trainer.register_callback(cfg)
+        trainer._register_hook(config.HOOK)
         for hook in trainer._hooks:
             if isinstance(hook, SchedulerHook):
-                assert True and len(trainer._hooks) == 2
+                assert True and len(trainer._hooks) == 1
```
Comment (mousyball) [inline tests/trainer/trainer_test.py]: Remove `True`.

File: tests/trainer/trainer_test.py [hunk 16]
```diff
@@ -55,53 +146,152 @@ def test_regist_optimizer_hook(self, cfg):
...
                 return
             else:
                 continue
         assert False
```
Comment (mousyball) [inline tests/trainer/trainer_test.py]: assert `False` at the end?

File: tests/trainer/trainer_test.py [hunk 17]
```diff
@@ -55,53 +146,152 @@ def test_regist_optimizer_hook(self, cfg):
...
+
+        trainer._register_hook(config.HOOK)
         for hook in trainer._hooks:
             if isinstance(hook, CheckpointHook):
-                assert True and len(trainer._hooks) == 2
+                assert True and len(trainer._hooks) == 1
                 return
             else:
                 continue
         assert False
```
Comment (mousyball) [inline tests/trainer/trainer_test.py]: Remove `True`
Weird assertion at the end.
Use `for ... else ...` instead.

**Review by mousyball — COMMENTED** [review]
Fix them!

File: examples/train_cifar10.py (hunk 5 above)
Comment (JanLin0817) [reply examples/train_cifar10.py]: Naming was modified by replacing all

File: configs/pytorch_trainer/hook/logger/logger_hook.yaml (hunk 2 above)
Comment (JanLin0817) [reply configs/pytorch_trainer/hook/logger/logger_hook.yaml:6]: solved

File: configs/pytorch_trainer/hook/trainer_hook.yaml (hunk 1 above)
Comment (JanLin0817) [reply configs/pytorch_trainer/hook/trainer_hook.yaml:3]: solved

File: docs/priority.md (hunk 4 above)
Comment (JanLin0817) [reply docs/priority.md:10]: Move to TODO

File: pytorch_trainer/trainer/base_trainer.py (hunk 7 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/base_trainer.py:58]: * handle folder slash
```python
osp.join(work_dir, '')[:-1]
```
* add time stamp
```python
+ f'_{date}'
```

File: pytorch_trainer/trainer/base_trainer.py (hunk 6 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/base_trainer.py]: solved

File: pytorch_trainer/trainer/base_trainer.py (hunk 8 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/base_trainer.py]: solved

File: pytorch_trainer/trainer/hooks/logger/base_logger.py (hunk 9 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/hooks/logger/base_logger.py]: solved

File: pytorch_trainer/trainer/hooks/logger/loss_meter.py (hunk 10 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/hooks/logger/loss_meter.py]: solved

File: pytorch_trainer/trainer/hooks/logger/tensorboard.py (hunk 11 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/hooks/logger/tensorboard.py:45]: * add this issue to todo list: rename {before/after}_{train/val}_batch
 * https://github.com/mousyball/pytorch-trainer/blob/trainer/pytorch_trainer/trainer/hooks/base_hook.py#L22

File: tests/trainer/hook_test.py (hunk 12 above)
Comment (JanLin0817) [reply tests/trainer/hook_test.py]: Add to todo list

File: tests/trainer/trainer_test.py (hunk 13 above)
Comment (JanLin0817) [reply tests/trainer/trainer_test.py:36]: Comment added

File: tests/trainer/trainer_test.py (hunk 14 above)
Comment (JanLin0817) [reply tests/trainer/trainer_test.py]: Comment added

File: tests/trainer/trainer_test.py (hunk 15 above)
Comment (JanLin0817) [reply tests/trainer/trainer_test.py]: solved

File: tests/trainer/trainer_test.py (hunk 16 above)
Comment (JanLin0817) [reply tests/trainer/trainer_test.py]: Comment added and solved

File: tests/trainer/trainer_test.py (hunk 17 above)
Comment (JanLin0817) [reply tests/trainer/trainer_test.py]: solved

**Review by mousyball — COMMENTED** [review]
LGTM
